In [13]:
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.svm import SVC

X_train = np.load('../../data/X_train.npy')
X_test = np.load('../../data/X_test.npy')
y_train = np.load('../../data/Y_train.npy')
y_test = np.load('../../data/Y_test.npy')

print(X_train.shape)

(560, 13)


# **Classificação**

Nosso objetivo aqui é montar um modelo de classificação, o qual classificará os slientes em confiáveis e não confiáveis.

A variável **default** é utilizada como proxy para o risco de inadimplência, permitindo classificar clientes em confiáveis e não confiáveis com base em padrões históricos de comportamento financeiro.

Testaremos dois modelos diferentes, sendo eles:
* **Randon Forest:** combina múltiplas árvores de decisão independentes, utilizando um sistema de votação para reduzir erros individuais e aumentar a robustez do modelo.

* **Gradient Boosting:** constrói árvores de forma sequencial, onde cada novo modelo busca corrigir os erros dos anteriores, resultando em maior capacidade preditiva.

* **Support Vector Machine:** O Support Vector Machine busca encontrar a melhor fronteira de separação entre classes, maximizando a margem entre clientes confiáveis e não confiáveis.

## **Randon Forest**

In [7]:
#modelo
model_rf = RandomForestClassifier()
model_rf.fit(X_train, y_train)

#previsao
y_pred_rf = model_rf.predict(X_test)

#avaliação
print("Resultados modleoRandom Forest")
report = classification_report(y_test, y_pred_rf, output_dict=True)

df_report = pd.DataFrame(report).transpose()
df_report

Resultados modleoRandom Forest


,precision,recall,f1-score,support
0.0,0.862385,0.912621,0.886792,103.000000
1.0,0.709677,0.594595,0.647059,37.000000
accuracy,0.828571,0.828571,0.828571,0.828571
macro avg,0.786031,0.753608,0.766926,140.000000
weighted avg,0.822027,0.828571,0.823434,140.000000


## **Gradient Boosting**

In [9]:
# modelo
model_gb = GradientBoostingClassifier()
model_gb.fit(X_train, y_train)

# previsão
y_pred_gb = model_gb.predict(X_test)

# avaliação
print("Resultados modelo Gradient Boosting")

report_gb = classification_report(y_test, y_pred_gb, output_dict=True)
df_report_gb = pd.DataFrame(report_gb).transpose()

df_report_gb

Resultados modelo Gradient Boosting


,precision,recall,f1-score,support
0.0,0.884615,0.893204,0.888889,103.000000
1.0,0.694444,0.675676,0.684932,37.000000
accuracy,0.835714,0.835714,0.835714,0.835714
macro avg,0.789530,0.784440,0.786910,140.000000
weighted avg,0.834356,0.835714,0.834986,140.000000


# **Support Vector Machine**

In [14]:
# modelo
model_svm = SVC(probability=True) # probability=True permite extrair probabilidades depois se precisar
model_svm.fit(X_train, y_train)

# previsão
y_pred_svm = model_svm.predict(X_test)

# avaliação
print("Resultados modelo SVM")

report_svm = classification_report(y_test, y_pred_svm, output_dict=True)
df_report_svm = pd.DataFrame(report_svm).transpose()

df_report_svm

Resultados modelo SVM


,precision,recall,f1-score,support
0.0,0.813559,0.932039,0.868778,103.000000
1.0,0.681818,0.405405,0.508475,37.000000
accuracy,0.792857,0.792857,0.792857,0.792857
macro avg,0.747689,0.668722,0.688626,140.000000
weighted avg,0.778742,0.792857,0.773555,140.000000


Agora que temos nosso modelos e suas respectivas métricas de avaliação, precisamos entender cada uma e, assim, decidir qual modelo usar. 
* De modo geral, a **precisão** mede a qualidade da nossa decisão (não queremos ser injustos com bons clientes), enquanto o **recall** mede o nosso poder de detecção (não queremos deixar caloteiros passarem despercebidos).


## **Escolha**

Considerando os resultados obtidos, observamos que o **Support Vector Machine (SVM)** apresentou o desempenho inferior na classificação da **Categoria 1** (clientes não confiáveis/inadimplentes). Devido a essa baixa performance em identificar o grupo de risco, o modelo foi descartado da seleção final.

A disputa concentrou-se entre o **Random Forest** e o **Gradient Boosting**. Ambos apresentaram níveis satisfatórios e muito similares de acurácia e precisão. No entanto, o Gradient Boosting destacou-se por obter um **Recall** superior para a categoria de risco (Categoria 1).

Enquanto o Random Forest identificou apenas **59%** dos casos reais de inadimplência, o Gradient Boosting alcançou **67%**. Em um cenário de análise de dados bancários, o objetivo primordial é mitigar o risco de crédito, evitando a concessão de empréstimos a perfis não confiáveis. Portanto, priorizamos o modelo com maior capacidade de detecção (maior Recall).

Com base nesses critérios, o **Gradient Boosting** foi o modelo escolhido para a implementação.